Implement ``detect_errors`` attribute on command of tool XML.

jmchilton opened the pull request:
If present, it can be one of
- `default`, no-op fallback to stdio tags and erroring on standard error output.
- `exit_code`, error if tool exit code is not 0. (The @jmchilton recommendation).
- `aggressive`, error if tool exit code is not 0 or either Exception: or Error: appears in standard error/output. (The @bgruening recommendation).

The `stdio` tag is a very heavy solution when many (most?) tools use either this `exit_code` or `aggressive` strategy leading to a lot of duplicated XML - and if there is one thing I like less than XML it is duplicated XML.

Refactoring and unit/functional tests to support and demonstrate this.

Run relevant functional tests with:

```
./run_tests.sh -framework -id detect_errors_aggressive
```

Run relevant unit tests:

```
nosetests test/unit/tools/test_parsing.py
```

jmchilton committed:
Implement detect_errors attribute on XML.

If present, it can be one of

 - "default", no-op fallback to stdio tags and erroring on standard error output.
 - "exit_code", error if tool exit code is not 0. (The @jmchilton recommendation).
 - "aggressive", error if tool exit code is not 0 or either Exception: or Error: appears in standard error/output. (The @bgruening recommendation).

Refactoring and unit/functional tests to support and demonstrate this.

Run functional test with:

    ./run_tests.sh -framework -id detect_errors_aggressive

Run relevant unit tests:

    nosetests test/unit/tools/test_parsing.py

Updated from original version to reflect comments on pull request #117 - in particular the ``detect_errors`` tag was moved from ``tool`` to ``command``.

nsoranzo commented:
What about making it an attribute of `<command>` instead of `<tool>`, since there can be only 1 command?

jmchilton commented:
@nsoranzo Is there a reason you don't like `<tool>`? `<tool>`seems to be more consistent with `<stdio>` tag which is a child of `<tool>` and not `<command>`?

hexylena commented:
@jmchilton IMO how exits are handled aren't tool metadata, they're command metadata. my gut prefers @nsoranzo's suggestion.

peterjc commented:
Aggressive should probably allow case variants like ERROR as well as Error.

natefoo commented:
FWIW, there can only be one command tag now, but it'd be handy to be able to be able to define more in the future for things like map/reduce-type workflows in a single tool, or embarrassingly parallel steps in between serial steps.

blankenberg commented:
+1 to the idea of potentially allowing multiple command tags. It might be nice to be able to put different failure modes on each command execution though.

jmchilton commented:
@peterjc I consider `aggressive` to be @bgruening's baby so I will let him weigh in on that modification. Whatever regex you guys decide is fine - but then we have to stick with it I think (post-release). We can always add other approaches though `sorta_aggressive`, `very_aggressive`, `check_java` (looks for standard Java style error logs and stack traces), `check_python` (checks for python tracebacks), etc... .

@blankenberg, @natefoo so I am going to assume you guys are siding with @erasche and @nsoranzo that the `detect_errors` should be on `command` then. If no one is with me that it is inconsistent with `stdio` I will make this modification.

bgruening commented:
@jmchilton I'm fine with @peterjc suggestion to make it case insensitive.
For the other question I'm on your side this is inconsistent, but this this does not mean it's correct in first place. If we are really moving to multiple commands, a great thing to do, we should start now to put this as attribute to the command tag.

jmchilton commented:
Okay - I will make these two modifications then.

Thanks for the input all!

jmchilton commented:
Okay - rebased and moved the tag and learned that these expressions have been case insensitive all along (though I made them lower case and added a comment to reflect this).

nsoranzo commented:
I think the title of this PR should be renamed `s/tool/command/` for people looking at it in the future.

jmchilton commented:
Done.

peterjc commented:
Other than the typo, commamd [sic] --> command :(
